core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -232,11 +237,21 @@ public ClusterState execute(ClusterState currentState) {
                     // Store newSnapshot here to be processed in clusterStateProcessed
                     List<String> indices = Arrays.asList(indexNameExpressionResolver.concreteIndexNames(currentState, request.indicesOptions(), request.indices()));
                     logger.trace("[{}][{}] creating snapshot for indices [{}]", repositoryName, snapshotName, indices);
+                    List<IndexId> snapshotIndices = new ArrayList<>();
+                    for (String index : indices) {
+                        final IndexId indexId;
+                        if (indexIds.containsKey(index)) {
+                            indexId = indexIds.get(index).getIndexId();
+                        } else {
+                            indexId = new IndexId(index, UUIDs.randomBase64UUID());
+                        }
+                        snapshotIndices.add(indexId);
+                    }
```

Comment:
I think this loop might land nicely as a ReposityData method `resolveIndices` or something like this.